src/Symfony/Security/State/IsGrantedAccessCheckerProvider.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/*
+ * This file is part of the API Platform project.
+ *
+ * (c) Kévin Dunglas <dunglas@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Symfony\Security\State;
+
+use ApiPlatform\Metadata\Operation;
+use ApiPlatform\State\ProviderInterface;
+use ApiPlatform\Symfony\Security\Exception\AccessDeniedException;
+use Symfony\Component\Security\Core\Authorization\AuthorizationCheckerInterface;
+use Symfony\Component\Security\Http\Attribute\IsGranted;
+
+/**
+ * Denies access based on the ApiPlatform\Metadata\ResourceAccessCheckerInterface.
+ * This implementation covers HTTP.
+ *
+ * @see ResourceAccessCheckerInterface
+ */
+final class IsGrantedAccessCheckerProvider implements ProviderInterface
+{
+    public function __construct(private readonly ProviderInterface $decorated, private readonly ?AuthorizationCheckerInterface $authChecker = null)
+    {
+    }
+
+    public function provide(Operation $operation, array $uriVariables = [], array $context = []): object|array|null
+    {
+        if (!$this->authChecker) {
+            return $this->decorated->provide($operation, $uriVariables, $context);
+        }
+
+        $s = $operation->getSecurity();
+        if (null === $s) {
+            return $this->decorated->provide($operation, $uriVariables, $context);
+        }
+
+        if (!\is_array($s)) {
+            $s = [$s];
+        }
+
+        foreach ($s as $isGranted) {
+            if (!$isGranted instanceof IsGranted) {
+                continue;
+            }
+
+            if (!$this->authChecker->isGranted($isGranted->attribute, $isGranted->subject)) {
```

Comment:
I wonder if we shouldn't just add this logic in `ResourceAccessChecker::isGranted()`.

As this class is final, we can change add types to the `expression` parameter if needed.

This will simplify a bit the code, and prevent an extra calls to several methods (`$operation->getSecurity()`).

Reply:
We still need a provider that runs before the read provider but I could use the same provider is that what you're saying? Then, in this first call we'd skip all expression that are not `IsGranted` attributes? It's probably slightly less code but the logic of the ResourceAccessChecker would be not the same according to when it is called (which is weird in my opinion) for example if you've the `object` then we must not call the given security.